src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs

```diff
@@ -1147,5 +1150,111 @@ static void Main(object[] args)
                 //                 Foo();
                 Diagnostic(ErrorCode.ERR_UseDefViolation, "Foo()").WithArguments("x").WithLocation(9, 17));
         }
+
+        [Fact]
+        [WorkItem(14097, "https://github.com/dotnet/roslyn/issues/14097")]
+        public void PiecewiseStructAssign()
+        {
+            var comp = CreateCompilationWithMscorlib(@"
+struct S { public int X, Y; }
+
+class C
+{
+    public static void Main()
+    {
+        S s;
+        s.X = 5;
+        void Local()
+        {
+          s.Y = 10;
+          System.Console.WriteLine(s);
+        }
+        Local();
+    }
+}");
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(14097, "https://github.com/dotnet/roslyn/issues/14097")]
+        public void PiecewiseStructAssign2()
+        {
+            var comp = CreateCompilationWithMscorlib(@"
+struct S
+{
+    public int X;
+    public int Y { get; set; }
+    
+    public S(int x, int y)
+    {
+        this.X = x;
+        this.Y = y;
+
+        Local(this);
+        void Local(S s)
+        {
+            s.X++;
+            s.Y++;
+        }
+    }
+}");
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(14097, "https://github.com/dotnet/roslyn/issues/14097")]
+        public void PiecewiseStructAssign3()
+        {
+            var comp = CreateCompilationWithMscorlib(@"
+struct S { }
+struct S2
+{ 
+    public int x;
+    public S s;
+}
+class C
+{
+    public void M()
+    {
+        S2 s2;
+        void Local()
+        {
+            s2.x = 0;
+            S2 s4 = s2;
+        }
+        Local();
+        S2 s3 = s2;
+    }
+
+    public void M2()
+    {
+        S2 s3;
+        void Local()
+        {
+            s3.s = new S();
+            S2 s4 = s3;
+        }
+        Local();
+    }
+
+    public void M3()
+    {
+        S2 s5;
+        void Local()
+        {
+            s5.s = new S();
+        }
+        Local();
+        S2 s6 = s5;
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (30,9): error CS0170: Use of possibly unassigned field 'x'
+                //         Local();
```

Comment:
>Local [](start = 27, length = 5)

Consider using different names for local functions, it will be easier to map errors to scenarios.